### PR TITLE
Bump `ginkgolinter` to `v0.13.3`

### LINKED ISSUE #10120: Introduce a linter to disallow ginkgo `FDescribe`, `FIt`, `FContext`, etc.
**Is your feature request related to a problem? Please describe**:
While developing locally, it's usually valuable to change ginkgo's `It`, `Context`, `Describe` to `FIt`, `FContext`, `FDescribe`, etc. This addition is being used to focus on certain function tests when running the test suite locally.

However, sometimes developers forget to remove the `F` prefix before submitting a PR. This can be discovered during the code-review process (as happened here: https://github.com/kubevirt/kubevirt/pull/10099#discussion_r1263159334), or even worse - the PR can get merged demanding another fix PR to remove the prefix.

**Describe the solution you'd like**:
I'd like to see a linter that would run as a gating presubmit lane on every PR, that would fail such PRs from getting merged.
Possibly the best way to implement such a linter is via `nogo`.

**Describe alternatives you've considered**:
Open to hear if they exist :)

/label good-first-issue

### PR DESCRIPTION
**Pull request by lyarwood**
**What this PR does / why we need it**:

#10120 had asked for a linter to be introduced to avoid focus containers from being committed or merged into the project. Such a check was added into `ginkgolinter` in version `v0.13.3`. This PR first bumps `ginkgolinter` to `v0.13.3` and then enables the focus container check in CI.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #10120

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nunnatsa** [comment]
A few comments on the PR description:

> ... being unable to use the focus containers when testing functional tests locally or in CI.

I think we should not allow focused containers in CI. Also, running unit tests using `go test` does not use nogo, and so it is still possible even if the new rule will be enabled.

What I'm not sure about, is if the scenario of running tests with bazel (and so with nogo) is an actual thing, when the developer want to focus in a specific test(s) (and so they will use focused containers). 

If we can assume that this is not a real use case, I think we should consider to enable the new linter rule.

@iholder101, @enp0s3 - WDYT? please also cc anyone who can answer these questions.

**Review by 0xFelix — APPROVED** [review]
/lgtm

**Comment by lyarwood** [comment]
> A few comments on the PR description:
> 
> > ... being unable to use the focus containers when testing functional tests locally or in CI.
> 
> I think we should not allow focused containers in CI. Also, running unit tests using `go test` does not use nogo, and so it is still possible even if the new rule will be enabled.

I disagree, I think it's entirely valid for a dev working on a draft (or not) PR to use focused containers to validate a specific functional test on a specific test lane that might provide an environment they don't have access to locally.

> What I'm not sure about, is if the scenario of running tests with bazel (and so with nogo) is an actual thing, when the developer want to focus in a specific test(s) (and so they will use focused containers).

It's used during our functional test runs, left enabled the following workflow is no longer possible locally or in CI:

```
$ git reset --hard HEAD~1
$ git diff
# git diff
diff --git a/tests/instancetype_test.go b/tests/instancetype_test.go
index 4140f9e61..f340cff84 100644
--- a/tests/instancetype_test.go
+++ b/tests/instancetype_test.go
@@ -41,7 +41,7 @@ import (
        "kubevirt.io/kubevirt/tests/util"
 )
 
-var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-compute] Instancetype and Preferences", decorators.SigCompute, func() {
+var _ = FDescribe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-compute] Instancetype and Preferences", decorators.SigCompute, func() {
 
        var (
                virtClient kubecli.KubevirtClient
$ make functest
hack/dockerized "hack/bazel-fmt.sh && hack/bazel-build-functests.sh"
go version go1.19.9 linux/amd64

go version go1.19.9 linux/amd64
Sandbox is up to date
INFO: Analyzed target //vendor/mvdan.cc/sh/v3/cmd/shfmt:shfmt (6 packages loaded, 33 targets configured).
INFO: Found 1 target...
Target //vendor/mvdan.cc/sh/v3/cmd/shfmt:shfmt up-to-date:
  bazel-bin/vendor/mvdan.cc/sh/v3/cmd/shfmt/shfmt_/shfmt
INFO: Elapsed time: 0.865s, Critical Path: 0.75s
INFO: 12 processes: 1 internal, 11 processwrapper-sandbox.
INFO: Build completed successfully, 12 total actions
INFO: Build completed successfully, 12 total actions
INFO: Analyzed target //:gazelle (0 packages loaded, 438 targets configured).
INFO: Found 1 target...
Target //:gazelle up-to-date:
  bazel-bin/gazelle-runner.bash
  bazel-bin/gazelle
INFO: Elapsed time: 0.192s, Critical Path: 0.03s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
INFO: Analyzed target //:goimports (0 packages loaded, 44 targets configured).
INFO: Found 1 target...
Target //:goimports up-to-date:
  bazel-bin/goimports.bash
INFO: Elapsed time: 0.192s, Critical Path: 0.03s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
INFO: Analyzed target //:buildifier (0 packages loaded, 291 targets configured).
INFO: Found 1 target...
Target //:buildifier up-to-date:
  bazel-bin/buildifier.bash
INFO: Elapsed time: 0.166s, Critical Path: 0.03s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
Sandbox is up to date
DEBUG: /root/.cache/bazel/_bazel_root/6f347497f91c9a385dcd9294645b76e0/external/bazel_gazelle/internal/go_repository.bzl:209:18: org_golang_x_net: gazelle: finding module path for import golang.org/x/sys/windows: finding module path for import golang.org/x/sys/windows: package golang.org/x/sys/windows: build constraints exclude all Go files in /root/.cache/bazel/_bazel_root/6f347497f91c9a385dcd9294645b76e0/external/bazel_gazelle_go_repository_cache/pkg/mod/golang.org/x/sys@v0.10.0/windows
gazelle: finding module path for import golang.org/x/sys/windows: finding module path for import golang.org/x/sys/windows: package golang.org/x/sys/windows: build constraints exclude all Go files in /root/.cache/bazel/_bazel_root/6f347497f91c9a385dcd9294645b76e0/external/bazel_gazelle_go_repository_cache/pkg/mod/golang.org/x/sys@v0.10.0/windows
gazelle: finding module path for import golang.org/x/sys/windows: finding module path for import golang.org/x/sys/windows: package golang.org/x/sys/windows: build constraints exclude all Go files in /root/.cache/bazel/_bazel_root/6f347497f91c9a385dcd9294645b76e0/external/bazel_gazelle_go_repository_cache/pkg/mod/golang.org/x/sys@v0.10.0/windows
INFO: Analyzed 7 targets (0 packages loaded, 4026 targets configured).
INFO: Found 7 targets...
ERROR: /root/go/src/kubevirt.io/kubevirt/tests/BUILD.bazel:107:8: GoCompilePkg tests/go_default_test_test.external.a failed: (Exit 1): builder failed: error executing command bazel-out/k8-opt-exec-2B5CBBC6/bin/external/go_sdk/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -tags selinux,selinux -src tests/io_utils.go -src tests/job.go -src ... (remaining 419 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
compilepkg: nogo: errors found by nogo during build-time code analysis:
tests/instancetype_test.go:44:9: ginkgo-linter: Focus container found. This is used only for local debug and should not be part of the actual source code, consider to replace with "Describe" (ginkgolinter)
INFO: Elapsed time: 26.317s, Critical Path: 23.18s
INFO: 1140 processes: 2 internal, 1138 processwrapper-sandbox.
FAILED: Build did NOT complete successfully
make: *** [Makefile:19: bazel-build-functests] Error 1
```

**Comment by nunnatsa** [comment]
The issue in func test is much easier to solve, using ginkgo command line flags.

**Comment by lyarwood** [comment]
> The issue in func test is much easier to solve, using ginkgo command line flags.

I'm not sure how `ginkgo` flags will help when we can't even build the functional tests?

The `// ginkgo-linter:ignore-focus-container-warning` comment will workaround this but it's pretty awkward.

**Comment by EdDev** [comment]
> developers being unable to use the focus containers when testing functional tests locally or in CI.

@lyarwood , this is surely valid for running tests locally, but I somehow remember that focus tests are ignored on CI.

> The // ginkgo-linter:ignore-focus-container-warning comment will workaround this but it's pretty awkward.

I agree, commenting exceptions in-code is bad.

> 
https://github.com/kubevirt/kubevirt/pull/10002 is seeking to introduce a pre-commit config using ginkgolinter that would however be capable of using this check and could be easily overridden in cases where a dev wanted to push a commit to CI for testing.

Right, this is a good use-case in favor of the separate make target.
Hope you'll manage to convince a maintainer to take it in.

File: nogo_config.json
Comment (EdDev) [inline nogo_config.json:145]: Can this be override using a command argument?
I suspect CI is ignoring focus tests anyway, so a job definition or the CI entry point script could use a flag/argument to control it.

**Review by EdDev — APPROVED** [review]

**Comment by 0xFelix** [comment]
@iholder101 Or WDYT about introducing a new target for CI (e.g. `make functest-ci` or `make functest-strict` or `make functest-no-focus`) while keeping the current behavior on the existing `make functest`?

**Comment by lyarwood** [comment]
@iholder101 @0xFelix what about https://github.com/kubevirt/kubevirt/pull/10136/commits/b3edf0098fb9982a8d72c6d75a9d6529f8aaecf1 that now removes the flag when running in CI. This should allow devs to use their normal workflows locally and cause CI to fail during the build if the focus containers are still present when a change is pushed?

File: nogo_config.json
Comment (lyarwood) [reply nogo_config.json:145]: I wasn't able to find a way of passing these through to `bazel`, for now I've added a hackaround in b3edf0098fb9982a8d72c6d75a9d6529f8aaecf1 that modifies the `nogo_config.json` file directly when the `KUBEVIRT_E2E_FOCUS` env variable is set.

**Comment by 0xFelix** [comment]
I like it, that's even better.

/lgtm

File: hack/bazel-build-functests.sh
Comment (EdDev) [inline hack/bazel-build-functests.sh:28]: If the intention is to act based on CI, isn't `CI` the correct env var to check against?
This `*FOCUS` one has a total different purpose (filtering which tests to run).

There may be another problem with this change, but I am not sure if the effect is visible or just theoretical: The modified file is part of BAZEL config, i.e. one of its input, therefore, touching it may invalidate its cache and take out one of BAZEL main advantages. That cache is shared between running instances on CI.

IMO this is a nasty hack, I do not think it is worth it. Hacks usually bite back sooner or later.
There is an elegant solution to this, to run an out-of-bazel check as suggested originally.

**Review by EdDev — CHANGES_REQUESTED** [review]

**Comment by enp0s3** [comment]
Hi,

I think the following may work (need to test if first):
1. Create additional nogo config file called `nogo_config_ci.json`, in that file enable the foucs linter.
2. in the root project BUILD.bazel define the following:
```
config_setting(
  name = "ci_env",
  values = {"execution_env": "ci"},
)
...
...
nogo(
    name = "nogo_vet",
    config = select({
      ":ci_env":  "nogo_config_ci.json",
      "//conditions:default": "nogo_config.json",
 })
...
...
)

```
3. Inject the configured flag `--execution_env=ci` into the `ci.bazelrc` file, which is being imported
only when in CI.

**Comment by enp0s3** [comment]
@lyarwood @0xFelix @EdDev If what I've mentioned will work, we can create a separate lane that will use ginkgo linter to catch focused tests.

**Comment by EdDev** [comment]
> @lyarwood @0xFelix @EdDev If what I've mentioned will work, we can create a separate lane that will use ginkgo linter to catch focused tests.

IMO it is an overkill. I think https://github.com/kubevirt/kubevirt/pull/10002 is enough and it has a job already which can catch it.
To me, all the other options that I see, including the one with BAZEL config you suggested, is a redundant complexity.
I guess it is better than the patching option of the config, as presented in t his PR now.

I would not recommend to create a job just to check this, it should be part of a generic linter one.
It just does not scale well, so I would try to find an existing job that this can fit under first.

**Comment by nunnatsa** [comment]
New lane for ginkgolinter may be overkill. But what about a new lane for golangci-lint, when ginkgolinter is enabled?

File: hack/bazel-build-functests.sh
Comment (enp0s3) [reply hack/bazel-build-functests.sh:28]: @lyarwood @EdDev Exactly, we have `"$CI"` which is set to `true` when running in CI env

**Comment by enp0s3** [comment]
> > @lyarwood @0xFelix @EdDev If what I've mentioned will work, we can create a separate lane that will use ginkgo linter to catch focused tests.
> 
> IMO it is an overkill. I think #10002 is enough and it has a job already which can catch it. To me, all the other options that I see, including the one with BAZEL config you suggested, is a redundant complexity. I guess it is better than the patching option of the config, as presented in t his PR now.
> 
> I would not recommend to create a job just to check this, it should be part of a generic linter one. It just does not scale well, so I would try to find an existing job that this can fit under first.

@EdDev Which job?

**Comment by enp0s3** [comment]
> > @lyarwood @0xFelix @EdDev If what I've mentioned will work, we can create a separate lane that will use ginkgo linter to catch focused tests.
> 
> IMO it is an overkill. I think #10002 is enough and it has a job already which can catch it. To me, all the other options that I see, including the one with BAZEL config you suggested, is a redundant complexity. I guess it is better than the patching option of the config, as presented in t his PR now.
> 
> I would not recommend to create a job just to check this, it should be part of a generic linter one. It just does not scale well, so I would try to find an existing job that this can fit under first.

@EdDev I see. I also think #10002 its a good solution but currently not all the test code is being covered in `make lint`

**Comment by EdDev** [comment]
> New lane for ginkgolinter may be overkill. But what about a new lane for golangci-lint, when ginkgolinter is enabled?

`golangci-lint` has a job already, but there are two points to resolve there:
- Make it "required".
- Run golangci-lint a second time only for this ginkgo linter for all the project source file (just change it in the Makefile target).

I think these are resolvable.

**Comment by nunnatsa** [comment]
> Run golangci-lint a second time only for this ginkgo linter for all the project source file (just change it in the Makefile target).

Why second time? Just enable ginkgolinter in .golangci.yml

**Comment by EdDev** [comment]
> > Run golangci-lint a second time only for this ginkgo linter for all the project source file (just change it in the Makefile target).
> 
> Why second time? Just enable ginkgolinter in .golangci.yml

Because only a few folders are covered currently.
If we want to run the linter for all sources with ginkgolinter enabled, we need to run in again.

**Comment by lyarwood** [comment]
/hold

After a quick discussion this morning the following path forward has been agreed:

* @nunnatsa will switch the behaviour of the  `suppress-focus-container` within ginkgolinter to opt-in once @EdDev has raised an issue
* @lyarwood will reach out to CI folks about bazel caching invalidation when updating the nogo config
* @lyarwood will continue to push https://github.com/kubevirt/kubevirt/pull/10002 forward, introducing a second run of `golangci-linter` running only `ginkgolinter` against everything (as already validated previously by nogo) before adding an optional `pre-commit` config

**Comment by nunnatsa** [comment]
> * @nunnatsa will switch the behaviour of the  `suppress-focus-container` within ginkgolinter to opt-in once @EdDev 

[ginkgolinter v0.13.3 is out](https://github.com/nunnatsa/ginkgolinter/releases/tag/v0.13.3)

The focused container rule is now disabled by default.

Use the `--forbid-focus-container=true` flag to enable it.

File: hack/bazel-build-functests.sh
Comment (lyarwood) [reply hack/bazel-build-functests.sh:28]: ACK forgot to set this in the latest run but I'll change it once the compute test has finished.

File: hack/bazel-build-functests.sh
Comment (nunnatsa) [inline hack/bazel-build-functests.sh:30]: I tried this line and it's working. I'm not sure what happen to the json colors when redirecting the jq output to a file. I would use the `-M` flag just for case. e.g.
```suggestion
    jq -M '.ginkgolinter.analyzer_flags |= . + {"forbid-focus-container":"true"}' nogo_config.json > nogo_config_ci.json && mv nogo_config_ci.json nogo_config.json
```

Comment (nunnatsa) [reply hack/bazel-build-functests.sh:30]: After reading about it, jq default is to use color only when writing to the terminal, but not for files. So this is not needed.

Comment (lyarwood) [reply hack/bazel-build-functests.sh:30]: Yeah the resulting diff looked clean to me but thanks for checking.

**Comment by lyarwood** [comment]
/hold cancel

@EdDev @nunnatsa @iholder101 @enp0s3 - I spoke to @brianmcarey about this and we think that aside from the first build and invalidation subsequent builds should use the cache as normal with this change in place. Moving this back to ready to review as a result.

**Review by 0xFelix — APPROVED** [review]
/lgtm

**Comment by iholder101** [comment]
@lyarwood So, to summarize this PR:
* Locally, developers would be able to run `make functest` with `F` prefix to focus on specific tests
* On CI, the build would fail if `F` prefix is found.

It this correct?

**Comment by lyarwood** [comment]
> @lyarwood So, to summarize this PR:
> 
> * Locally, developers would be able to run `make functest` with `F` prefix to focus on specific tests
> * On CI, the build would fail if `F` prefix is found.
> 
> It this correct?

Yes correct, I'd also note that in CI the `F` prefix does nothing anyway thus we are safe to block its use.

**Comment by iholder101** [comment]
> > @lyarwood So, to summarize this PR:
> > 
> > * Locally, developers would be able to run `make functest` with `F` prefix to focus on specific tests
> > * On CI, the build would fail if `F` prefix is found.
> > 
> > It this correct?
> 
> Yes correct, I'd also note that in CI the `F` prefix does nothing anyway thus we are safe to block its use.

Yup, agreed.
Great work @lyarwood and everyone else involved!

/approve

**Comment by iholder101** [comment]
@lyarwood would you add https://github.com/kubevirt/kubevirt/issues/10120 under `Fixes #` in the PR description to link this PR as a fix for the issue? Thanks!

**Comment by nunnatsa** [comment]
> Yes correct, I'd also note that in CI the `F` prefix does nothing anyway thus we are safe to block its use.

This is right but not accurate. This change will prevent leaking of focused containers into upstream, that will misguide other developer locally. The fact that this focused containers are still working at upstream just make it harder to find these leaks. This is why this chage is important.

**Comment by EdDev** [comment]
> /hold cancel
> 
> @EdDev @nunnatsa @iholder101 @enp0s3 - I spoke to @brianmcarey about this and we think that aside from the first build and invalidation subsequent builds should use the cache as normal with this change in place. Moving this back to ready to review as a result.

@brianmcarey , has this been confirmed in practice or is it just in theory?
I guess that now that is got merged, it should be easier to check.

**Pull request merged**
